src/mlib/path.h

```diff
@@ -0,0 +1,172 @@
+#ifndef MONGOCRYPT_PATH_PRIVATE_H
+#define MONGOCRYPT_PATH_PRIVATE_H
+
+#include "./user-check.h"
+
+#include "./str.h"
+
+#include <inttypes.h>
+
+/**
+ * @brief The preferred path separator for the current platform
+ */
+static const char MPATH_PREFERRED_PATH_SEPARATOR =
+#ifdef _WIN32
+   '\\'
+#else
+   '/'
+#endif
+   ;
+
+/**
+ * @brief Determine if the given character is a path separator on the current
+ * platform.
+ */
+static inline bool
+mpath_is_sep (char c)
+{
+#ifdef _WIN32
+   return c == '/' || c == '\\';
+#else
+   return c == '/';
+#endif
+}
+
+/**
+ * @brief Determine whethre the given path string has a trailing path separator
+ */
+static inline bool
+mpath_has_trailing_sep (mstr_view path)
+{
+   return path.len && mpath_is_sep (path.data[path.len - 1]);
+}
+
+/**
+ * @brief Obtain the parent path of the given path.
+ *
+ * @param path A path string
+ * @return mstr_view A substring of the given path string that views only the
+ * parent directory of the given path.
+ */
+static inline mstr_view
+mpath_parent (mstr_view path)
```

Comment:
iiuc, this means that `mpath_parent()` returns an empty string if the input is `/tmp` or (on win32) `C:\`. Is that expected? At least the path handling libraries of Python and Node.js differ in returning `/` and `C:\`, respectively, and I think that would be the most reasonable behavior here as well.

Does this code need to be able to handle Windows UNC paths here (e.g. `\\?\Ç:\Users\me`)? Since the PR description says that this is a general-purpose library and it will handle user-defined input, I would assume that the answer is ‘yes’ as well.

(Also, speaking from Python and Node.js experience, both platforms also provide Windows-specific and POSIX-specific versions of their path handling functions regardless, and will only switch the default depending on the host OS. That has helped a lot with testing, since it means that developers can work on the path handling code of one platform without necessarily being on that platform.)